manifests/init.pp

```diff
@@ -53,5 +63,14 @@
     }
   }
 
-  create_resources('static_custom_facts::fact', $custom_facts)
+  $custom_facts.each |$name, $value| {
+    $ensure = defined('$value') ? {
+      true    => 'present',
+      default => 'absent',
+    }
+    static_custom_facts::fact { $name:
+      ensure => $ensure,
+      value  => $value,
+    }
+  }
```

Comment:
Is this change effectively to better manage the value of `ensure`, or are you trying to do something else here that `create_resources` can't do for us?

Reply:
Both. While create_resources will basically do the job, the resources created are not contained in any other class making it hard to manage dependencies and ordering. Also, debugging resource conflicts is much harder if you cannot exactly know were the conflicting resources in the catalog came from. A factory pattern, as implemented here using the each iterator, will create resources all contained within the class static_custom_facts and that will be clear when looking on the dependency graph of the catalog.

The piece of magic with the ensure allows you to drop certain custom facts by defining them with undef value in hiera, without the need to enable purge_unmanaged. By enabling a fine grained handling of custom facts, this takes away some pain if we want to move from custom facts already existing to get them handled by this module (what is exactly what we're striving for).